### PR TITLE
Decrypt encrypted attachment

### LINKED ISSUE #195: decrypt encrypted attachments during download
part of  #275, after #194

### PR DESCRIPTION
**Pull request by ekievsky**
This PR adds file decryption of email attachments

close #195

----------------------------------

**Tests**:
- Tests added or updated

--------------------------------

### To be filled by reviewers

I have reviewed that this PR... _(tick whichever items you personally focused on during this review)_:
- [x] addresses the issue it closes (if any)
- [x] code is readable and understandable
- [x] is accompanied with tests, or tests are not needed
- [x] is free of vulnerabilities
- [x] is documented clearly and usefully, or doesn't need documentation

### CODE REVIEW

**Comment by ekievsky** [comment]
@tomholub which task should I take next?

File: FlowCrypt/Core/Core.swift [hunk 1]
```diff
@@ -21,6 +23,8 @@ enum CoreError: Error, Equatable {
         switch errorType {
         case "format": self = .format(coreError.error.message)
         case "key_mismatch": self = .keyMismatch(coreError.error.message)
+        case "no_mdc": self = .noMDC(coreError.error.message)
+        case "bad_mdc": self = .badMDC(coreError.error.message)
         default: return nil
```
Comment (tomholub) [inline FlowCrypt/Core/Core.swift:28]: Need to check in another PR that all errors are represented - at least it's missing need_passphrase and bad_passphrase

File: FlowCrypt/Functionality/Mail Provider/Message Provider/MessageService.swift [hunk 2]
```diff
@@ -125,18 +130,29 @@ final class MessageService {
...
             .filter { $0.decryptErr != nil }
 
-        let attachments = decrypted.blocks
+        let attachments = try decrypted.blocks
             .filter(\.isAttachmentBlock)
-            .map(MessageAttachment.init)
+            .map { block -> MessageAttachment in
+                var name = block.attMeta?.name ?? "Attachment"
+                let size = block.attMeta?.length ?? 0
+                var data = block.attMeta?.data ?? Data()
```
Comment (tomholub) [inline FlowCrypt/Functionality/Mail Provider/Message Provider/MessageService.swift:147]: These three cannot should not be empty. If attMeta is missing, then it's not an attachment, and can be skipped. You can add a guard for that.

File: FlowCryptCommon/Extensions/StringExtensions.swift [hunk 3]
```diff
@@ -45,6 +45,20 @@ public extension String {
             }
         }
     }
+    
+    var fileExtension: String? {
+        guard let fileExtension = self.split(separator: ".").last else {
+            return nil
+        }
+        return String(fileExtension)
+    }
```
Comment (tomholub) [inline FlowCryptCommon/Extensions/StringExtensions.swift:54]: seems like a bug? file named just `pgp` will be seen as having extension pgp, which it doesn't.

File: FlowCryptCommon/Extensions/StringExtensions.swift [hunk 4]
```diff
@@ -45,6 +45,20 @@ public extension String {
...
+    
+    func dropExtension() -> String {
+        guard let fileExtension = self.fileExtension else {
+            return self
+        }
+        return self.replacingOccurrences(of: ".\(fileExtension)", with: "")
+    }
```
Comment (tomholub) [inline FlowCryptCommon/Extensions/StringExtensions.swift:61]: File named `made.by.pgpainless.pgp` would come out as `made.byainless` which is wrong. User a swift method to remove string suffix if there is one. Else a regex with a clear $ at the end.

File: FlowCrypt/Functionality/Mail Provider/Message Provider/MessageService.swift [hunk 5]
```diff
@@ -125,18 +130,29 @@ final class MessageService {
...
+
+                if name.fileExtension == Constants.encryptedAttachmentExtension {
+                    data = (try core.decryptFile(encrypted: data, keys: keys, msgPwd: nil).content)
+                    name = name.dropExtension()
+                }
```
Comment (tomholub) [inline FlowCrypt/Functionality/Mail Provider/Message Provider/MessageService.swift:152]: Just realized, please look closely at the BlockType. If it's returned as `encryptedFile` then you decrypt it, else leave as is. This means you don't have to check the file name at all.

https://github.com/FlowCrypt/flowcrypt-ios/blob/ac29a9c92908d0d71c67c53b099080ecfcae5837/FlowCrypt/Core/CoreTypes.swift#L200

**Review by tomholub — COMMENTED** [review]
Some code comments - haven't tested yet.

File: FlowCrypt/Core/Core.swift (hunk 1 above)
Comment (ekievsky) [reply FlowCrypt/Core/Core.swift:28]: It's already handled when user opens message screen, but I may add it anyway

File: FlowCrypt/Functionality/Mail Provider/Message Provider/MessageService.swift (hunk 2 above)
Comment (ekievsky) [reply FlowCrypt/Functionality/Mail Provider/Message Provider/MessageService.swift:147]: just copy pasted init of MessageAttachment

File: FlowCryptCommon/Extensions/StringExtensions.swift (hunk 3 above)
Comment (ekievsky) [reply FlowCryptCommon/Extensions/StringExtensions.swift:54]: removed as unused

File: FlowCryptCommon/Extensions/StringExtensions.swift (hunk 4 above)
Comment (ekievsky) [reply FlowCryptCommon/Extensions/StringExtensions.swift:61]: done

File: FlowCrypt/Functionality/Mail Provider/Message Provider/MessageService.swift (hunk 5 above)
Comment (ekievsky) [reply FlowCrypt/Functionality/Mail Provider/Message Provider/MessageService.swift:152]: done

**Review by tomholub — COMMENTED** [review]
code looks good. will test

**Review by tomholub — APPROVED** [review]
excellent, works very nicely

haven't tested error scenarios

**Pull request merged**